nemo_rl/environments/search/search.py

```diff
@@ -0,0 +1,340 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import re
+import time
+import uuid
+from typing import Any, Dict, List, Optional, Tuple, TypedDict
+
+import ray
+import requests
+import torch
+
+from nemo_rl.data.interfaces import LLMMessageLogType
+from nemo_rl.distributed.batched_data_dict import BatchedDataDict
+from nemo_rl.environments.interfaces import EnvironmentInterface, EnvironmentReturn
+from nemo_rl.environments.search.utils import compute_score, extract_solution
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+DEFAULT_TIMEOUT = 30
+MAX_RETRIES = 10
+INITIAL_RETRY_DELAY = 1
+
+
+class SearchEnvConfig(TypedDict):
+    num_workers: int = 1
+    topk: int = 3
+    search_url: str = "http://localhost:8000/retrieve"
+    timeout: float = DEFAULT_TIMEOUT
+    max_turns: int = 10
+
+
+class SearchEnvMetadata(TypedDict):
+    ground_truth: str
+    num_turns: int = 0
+
+
+class SearchWorkerResult(TypedDict):
+    observation: str
+    reward: float
+    done: bool
+    error: Optional[str]
+    metadata: SearchEnvMetadata
+    answer: Optional[str]
+
+
+def call_search_api(
```

Comment:
```suggestion
def _call_search_api(
```
This seems more like an internal helper function to me.